### PR TITLE
refactor(esp32): hoist fractional clock-divider solver to core (#3562)

### LINKED ISSUE #3562: Hoist solveI2sClockDivider() to shared location; apply to other drivers
## Context

Follow-up from FastLED#3526 Phase 2e. The modern I2S engine's peripheral impl (`src/platforms/esp/32/drivers/i2s/i2s_peripheral_esp32dev_esp.cpp.hpp`) now contains a `solveI2sClockDivider(target_hz, &N, &A, &B)` helper — adapted from Yves's `i2s_define_bit_patterns` — that solves the 80 MHz APB / (N + B/A) fractional divider for any target frequency.

## The generalisation

The same fractional-divider math applies to other classic-ESP32 peripherals that derive their clock from the APB reference via a `clkm_conf.clkm_div_{num,a,b}` triple. Candidates:

- **SPI0/1/2/3**: `spi_dev_t::clock` divider (different register layout but same math)
- **RMT4 (classic ESP32)**: `rmt_dev_t::conf1.div_cnt_ch{n}` — 8-bit-only divider, but for finer control the fractional path applies
- **I2S0** on classic ESP32 (currently unused by any modern driver but the arithmetic is identical to I2S1)
- Anything on **ESP32-S2/S3 / P4 / C-series** — different peripheral families but the pattern generalises

## Proposal

Hoist `solveI2sClockDivider()` to a shared `src/platforms/esp/32/` location (e.g., `platforms/esp/32/core/clock_divider.h`) with a generic signature:

\`\`\`cpp
namespace fl::platforms::esp32 {
  struct ClkDividerNAB { int N; int A; int B; };
  ClkDividerNAB solveFractionalDivider(u32 base_hz, u32 target_hz,
                                        u8 max_A = 63, u8 min_N = 2) FL_NO_EXCEPT;
}
\`\`\`

Then:
- I2S peripheral consumes it (replace inline impl).
- SPI peripherals that currently open-code their divider math consume it.
- Future drivers pick it up "for free".

## Acceptance

- [ ] Investigate: audit `src/platforms/esp/32/` for callers currently open-coding N/A/B math.
- [ ] If ≥ 2 open-coders found: hoist the helper to shared location.
- [ ] Update I2S impl to consume the shared helper.
- [ ] Update other identified callers.
- [ ] Regression test: verify the hoisted helper produces byte-identical output to Yves's original `i2s_define_bit_patterns` for the canonical WS2811 / WS2812 / WS2812B timings.

## Reference

Current impl: `src/platforms/esp/32/drivers/i2s/i2s_peripheral_esp32dev_esp.cpp.hpp` — search for `solveI2sClockDivider`.

Origin: Yves Bazin's `i2s_esp32dev.cpp.hpp:pgcd()` + `i2s_define_bit_patterns()` divider search loop (deleted at FastLED#3526 Phase 2e; algorithm preserved in the hoisted helper).

Related: FastLED#3561 (PARLIO variable-timing — will likely consume the same helper).

### PR DESCRIPTION
**Pull request by zackees**
## Summary

Closes #3562.

`solveI2sClockDivider()` lived inside the I2S peripheral impl, but the `80 MHz APB / (N + B/A)` math applies to every classic-ESP32 peripheral that takes a `clkm_div_{num,a,b}` triple — SPI, RMT4, I2S0.

Moved to `platforms/esp/32/core/clock_divider.h` as:

```cpp
namespace fl { namespace platforms { namespace esp32 {
  struct ClkDividerNAB { int N; int A; int B; };
  ClkDividerNAB solveFractionalDivider(u32 base_hz, u32 target_hz,
                                       int max_A = 63, int min_N = 2) FL_NO_EXCEPT;
}}}
```

Pure arithmetic, no ESP headers — so it compiles and unit-tests on the host.

### One deliberate design choice

The I2S call site **keeps its own 8 MHz fallback** instead of pushing it into the shared solver. "What frequency do we want when none is configured" is driver policy, not divider arithmetic — baking `N=10` into shared math would silently hand every future consumer an I2S-specific default. The solver treats a zero target as degenerate and clamps to `min_N` rather than dividing by zero.

## Verification — behaviour-preserving, checked two ways

**1. Equivalence test.** `tests/platforms/esp/32/core/clock_divider.cpp` keeps a *verbatim copy* of the pre-hoist function and asserts the new solver returns identical `(N, A, B)` across the band the I2S driver actually uses (100 kHz → 40 MHz). This matters because a refactor that quietly retuned the divider would **not** fail any build — it would just shift every driver's output clock. Plus coverage for hardware limits (`A ≤ 63`, `B < A`, `N ≥ min_N`), realised-frequency error, base-clock generalisation, and degenerate inputs.

**2. Binary comparison.** `esp32dev` Blink is **391543 bytes before and after** — byte-identical.

### Why that size comparison is trustworthy

It nearly wasn't. `compile_commands.json` for `esp32dev` shows **zero** TUs matching `i2s`, which looks like the changed code isn't in the build at all — making a byte-identical size meaningless.

It isn't excluded; the drivers are **unity-built**, so the individual file path never appears. I confirmed the header really is compiled with a negative control: a temporary `#error` in it **failed the build, reported from 8 unity TUs**. Then reverted and re-verified green.

Flagging that because "grep compile_commands.json for the filename" is the obvious way to check this and it gives the wrong answer here.

## Testing

- `bash test --cpp` — all pass, including the new `clock_divider` target
- `bash compile esp32dev --examples Blink` — succeeds, flash unchanged at 391543 bytes
- `bash lint` — green

## Follow-on

This only hoists the math and repoints I2S at it. Actually adopting it in SPI / RMT4 (the rest of #3562's "apply to other drivers") is left alone — each needs its own register marshalling and bench validation, and bundling that here would make an intentionally byte-identical refactor into a behavioural change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added/standardized fractional clock-divider solving for ESP32, including safe behavior for unspecified and extreme inputs.
* **Bug Fixes**
  * Improved divider robustness with explicit clamping for zero inputs and ratio saturation to avoid undefined math.
  * Enhanced output accuracy via a fractional search approach and precision corner-case handling.
* **Refactor**
  * Updated the ESP32 I2S clock-divider calculation to reuse the shared solver for consistent results.
* **Tests**
  * Added host-side tests validating exact expected dividers, accuracy bounds, constraints, scaling, and saturation/degenerate cases.

### CODE REVIEW

File: src/platforms/esp/32/core/clock_divider.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        out.A = 1;
+        out.B = 0;
+        return out;
+    }
+
+    const double f_target = static_cast<double>(target_hz);
+    const double f_base = static_cast<double>(base_hz);
+    int N = static_cast<int>(f_base / f_target);
+    if (N < min_N) N = min_N;
+    const double residual = (f_base / f_target) - static_cast<double>(N);
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/platforms/esp/32/core/clock_divider.h]: Valid — fixed in c171aec.

`base_hz / target_hz` does exceed `INT_MAX` for extreme u32 pairs: `4294967295 / 1` is ~4.29e9 against an `INT_MAX` of ~2.147e9, so `static_cast<int>` on the ratio was undefined.

Key detail in the fix: the saturation has to happen **before** the cast, not after. Clamping the result afterwards would be clamping a value that is already undefined — there is nothing well-defined left to inspect. So the check is on the `double` ratio:

```cpp
const double ratio = f_base / f_target;
if (ratio >= static_cast<double>(kMaxDivider)) {
    out.n = kMaxDivider; out.a = 1; out.b = 0;
    return out;
}
int N = static_cast<int>(ratio);
```

Not reachable from the I2S caller (it passes ≥ 8 MHz), but the whole point of this PR is that other drivers will call it with their own inputs, so it shouldn't carry an undefined corner.

One deliberate limitation: `kMaxDivider` is an **int-representability** bound, not a hardware one. The real fields are much narrower — I2S `clkm_div_num` is 8 bits, so anything below ~313 kHz already overflows the register. Adding that hardware cap would change results for existing low-frequency inputs, which doesn't belong in a refactor whose whole claim is byte-identical output. Worth its own issue if you want the solver to enforce per-peripheral field widths.

Test added: `oversized ratios stay representable` covers the saturating pair, a just-under-the-clamp normal solve, and asserts saturation never yields an unusable divider.

Validation: 274/274 pass, `bash lint` green, and esp32dev Blink is still **391543 bytes** — byte-identical claim holds after the fix.

File: src/platforms/esp/32/core/clock_divider.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+/// Reference clock for APB-derived peripherals on the classic ESP32.
+constexpr u32 kApbClockHz = 80000000u;
+
+/// Solved divider triple: `f_out = base_hz / (N + B/A)`.
+struct ClkDividerNAB {
+    int N;
+    int A;
+    int B;
+};
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/platforms/esp/32/core/clock_divider.h:39]: Agreed — done in c171aec. `N`/`A`/`B` → `n`/`a`/`b`, with the I2S consumer and the tests updated.

I checked the guideline before changing it, since #3562's own proposal spelled them uppercase and they're the datasheet's field names. `cpp-standards.md:305` mandates `mCamelCase`, but that's aimed at class members with behaviour — the exemplar for plain aggregates at line 406 is `struct CacheKey { const DiodeProfile* p; u32 version; int cct; }`, i.e. plain lowercase, no prefix. So lowercase is right here and your call matches the codebase.

Kept the uppercase spelling in the doc comments, where it maps to the register names (`clkm_conf.clkm_div_{num,a,b}`) and the `base_hz / (N + B/A)` formula, so the datasheet correspondence isn't lost:

```cpp
/// Solved divider triple: `f_out = base_hz / (n + b/a)`.
///
/// Lowercase members per the plain-aggregate convention; the uppercase
/// N/A/B in the formulas below is the register-datasheet spelling
/// (`clkm_conf.clkm_div_{num,a,b}`).
```

File: src/platforms/esp/32/core/clock_divider.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zackees) [reply src/platforms/esp/32/core/clock_divider.h]: Took you up on the follow-up offer — filed as #3745.

Worth noting the concrete failure mode, because it's worse than "unvalidated input": the I2S peripheral assigns the solved `n` straight into a **C bitfield**

```cpp
i2s->clkm_conf.clkm_div_num = div_N;   // narrow field, no range check
```

and an out-of-range assignment there **silently truncates**. At the 80 MHz APB reference a 100 kHz target solves to `n = 800`; against an 8-bit field that becomes 32, so the hardware would run at 2.5 MHz instead of 100 kHz — a 25× error with no warning, no assert, and for a clockless LED driver just malformed WS2812 timing.

The sharper framing of the defect: the solver already encodes the `A` constraint (`max_A = 63` is exactly a 6-bit field) but has no equivalent for `n`. One hardware limit is expressed, its sibling isn't.

Latent rather than live — today's only caller passes ≥ 8 MHz — but it becomes reachable precisely because this PR turns the solver into shared infrastructure for peripherals with different field widths.

I also flagged in the issue that the specific 8-bit/6-bit widths come from the TRM and the existing `A ≤ 63` limit, and could not be confirmed against `soc/i2s_struct.h` from this checkout, so they should be verified per peripheral before being encoded.

File: src/platforms/esp/32/core/clock_divider.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        best_A = 1;
+        best_B = 0;
+        ++N;
+    }
+
+    out.N = N;
+    out.A = best_A;
+    out.B = best_B;
+    return out;
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/platforms/esp/32/core/clock_divider.h:127]: Declining this one — I implemented it, and it collides with another project rule that is a hard ban.

I did the split: declaration in `clock_divider.h`, definition moved to a new `clock_divider.cpp.hpp`, registered in `core/_build.cpp.hpp`. Then `bash lint` failed:

```
[CppHppIncludesChecker] tests/platforms/esp/32/core/clock_divider.cpp:
  Line 11: #include "platforms/esp/32/core/clock_divider.cpp.hpp"
  - Including *.cpp.hpp files in tests is banned (hard ban, no opt-out).
    Include the public header instead.
```

The two rules are mutually exclusive for this file:

- `src/**/*.h` must not contain function definitions → definition goes in `.cpp.hpp`
- tests must not include `.cpp.hpp` → the host test can only include the `.h`

And on the host there is no third path: `clock_divider.cpp.hpp` is only reachable through `core/_build.cpp.hpp`, which is part of the **ESP32-only** unity build. A declaration-only header therefore leaves the host test with a symbol it can never link. Removing the test isn't an option either — being host-testable pure math is the entire reason #3562 asked for this hoist, and the test is what pins the extraction as behaviour-preserving.

There was also a measurable cost: taking the function out-of-line moved esp32dev Blink from **391543 → 391547 bytes**, because it stops being inlined at its single call site and gets emitted as a real symbol. Small, but it breaks the byte-identical property this PR is specifically claiming.

So the header keeps the `inline` definition. Reverted, re-verified: `bash lint` green, esp32dev back to **391543 bytes**.

If the `src/**/*.h` rule should bind here, the prerequisite is a way to compile a pure-math `src/` implementation into the **host** test target — otherwise every host-testable helper hits this same wall. Happy to file that as its own issue if it's worth pursuing.
